presto-native-execution/presto_cpp/main/operators/ShuffleInterface.h

```diff
@@ -43,46 +29,65 @@ class ShuffleInterface {
   /// @param success set to false to indicate aborted client.
   virtual void noMoreData(bool success) = 0;
 
+  /// Return true if all the data is finished writing and is ready to
+  /// to be read while noMoreData signals the shuffle service that there
+  /// is no more data to be writen.
+  virtual bool readyForRead() const = 0;
+};
+
+class ShuffleReadInterface {
+ public:
+  virtual ~ShuffleReadInterface() = default;
+
   /// Check by the reader to see if more blocks are available for this
   /// partition.
   virtual bool hasNext(int32_t partition) const = 0;
 
   /// Read the next block of data for this partition.
   /// @param success set to false to indicate aborted client.
   virtual velox::BufferPtr next(int32_t partition, bool success) = 0;
+};
 
-  /// Return true if all the data is finished writing and is ready to
-  /// to be read while noMoreData signals the shuffle service that there
-  /// is no more data to be writen.
-  virtual bool readyForRead() const = 0;
+class ShuffleInterfaceFactory {
+ public:
+  virtual std::shared_ptr<ShuffleReadInterface> createReadInterface(
+      const std::string& serializedShuffleInfo,
+      velox::memory::MemoryPool* pool) = 0;
+
+  virtual std::shared_ptr<ShuffleWriteInterface> createWriteInterface(
+      const std::string& serializedShuffleInfo,
+      velox::memory::MemoryPool* pool) = 0;
 
   /// Register ShuffleInterfaceFactory to its registry. It returns true if the
   /// registration is successful, false if a factory with the name already
   /// exists.
   /// This method is not thread safe.
   static bool registerFactory(
       const std::string& name,
-      ShuffleInterfaceFactory shuffleInterfaceFactory) {
-    std::unordered_map<std::string, ShuffleInterfaceFactory>& factoryMap =
-        factories();
-    return factoryMap.emplace(name, shuffleInterfaceFactory).second;
+      std::unique_ptr<ShuffleInterfaceFactory>&& shuffleInterfaceFactory) {
```

Comment:
We don't need && for unique_ptr as we can't make copy of unique_ptr so just remove &&

Reply:
Right. :D